Implement `BidsPartialComponent`

pvandyken opened the pull request:
Resolves #243.

For a play-by-play of what I did, the commits are quite detailed and fairly well organized.

For a high level overview:

## The new classes

`BidsComponent` is conceived of as a multi-selectable `dict` (sort of... actually right now only `__getitem__` is implemented). If a tuple of entities is selected, you get back a `BidsPartialComponent` which has a `ZipList` but no `path` or `name`. `BidsComponent` is actually just a subclass of `BidsPartialComponent`. 

If a single entity is selected from `BidsComponent` (or `BidsPartialComponent`), you get a `BidsComponentRow`. This represents a single entity and its values, and it's implemented as a `ImmutableList` (sort of a `UserTuple`). It has the `wildcards` and `entities` properties, but instead of returning dicts, they directly return the `{str}` or `['list', 'of', 'values']` associated with the entity. This is to ensure symmetry between the following access modes:

```py
inputs['bold']['subject'].wildcards == inputs['bold'].wildcards['subject']
```

Note that the preferred style would still be `inputs['bold'].wildcards['subject']`, so this aspect of the new implementation is largely for consistency and to make everything more fail-proof.

On the other hand, `BidsComponentRow.zip_lists` *does* still return a dict mapping the one entity to its ordered entity list. This break in symmetry is motivated by the new `Expandable` protocol I established:

`Expandable` classes have an `expand` method, a `filter` method, and a `zip_lists` property or attribute. The signatures of these items must be as found in `BidsPartialComponent`. Thus, `zip_lists` in `BidsComponentRow` needs to return a dict in order to meet the protocol. In return, we're able to cleanly test all of three of the current classes using the exact same test code.

Furthermore, because `BidsComponentRow` is already implemented as a list with immediate access to the values, any call from the user to a symmetrically defined `zip_lists` would be completely redundant. It would give only a less powerful version of the list-like object the user already has.

## The documentation

I broke off a bit from our completely autogenerated documentation in the api section in order to prioritize the overview of `BidsComponent`. `BidsPartialComponent` is really mostly a repeat of `BidsComponent`, so none of it's members are shown, just an extended docstring overview. `BidsComponentRow` is rather different, so it's displayed in full.

I also moved the `input_*` legacy properties into a dropdown box. I suppose we could technically deprecate them, but they don't really do any harm and dropping them would completely prevent older workflows from upgrading (unlike `BidsDataset.wildcards` etc, which probably never had much buy-in). This way we can document them and keep them, but out of the way.

## Deduplication

I decided that `zip_list` deduplication should *not* be performed immediately when getting subcomponents and rows. This is both for simplicity, and to ensure that indexing remains consistent and clear as capabilities become more complex. So the only deduplication to be performed will be done in the `expand` method (since it unambiguously doesn't make sense for `expand` to return multiple identical paths)

## Design Question

* I stored the name of the entity in `BidsComponentRow` as `.entity`. The name is rather close to the `.entities` property, which returns something completely different (the `.entities` name doesn't make sense in isolation, but does if you consider the whole picture). So should we store `.entity` as something else?


## TBD

- [x] Add `zip_list` deduplication to the `expand` methods

pvandyken commented:
Okay, I think that finally fixes all the test failures.

Just note for posterity, when combining `MockerFixture` with a hypothesis test, you need to call `mocker.stopall()` at the very beginning of the test so that all the mocks are removed, otherwise you get very strange errors. This is even if you only run one example, because if the example fails, hypothesis still does multiple runs of the test to try to shrink the example.

tkkuehn reviewed (changes requested):
The code basically looks good to me, just a few changes for the docstrings (some of which are a matter of opinion and optional)

tkkuehn commented on snakebids/utils/utils.py:
```diff
@@ -356,4 +366,105 @@ def get_values(zlist: types.ZipListLike):
+        return self._data.index(value, start, stop)
+
+
+def get_wildcard_dict(__entities: str | Iterable[str]) -> dict[str, str]:
```
Here's the answer, for future reference: https://peps.python.org/pep-0484/#positional-only-arguments

pvandyken commented on snakebids/utils/utils.py:
```diff
@@ -1,21 +1,33 @@
 from snakebids.utils.user_property import UserProperty
 
-T = TypeVar("T")
+_T = TypeVar("_T")
```
Correct, it's the convention followed in all the typeshed annotations for built-in python modules, and I've organically started following it lately

pvandyken commented on snakebids/core/datasets.py:
```diff
@@ -36,18 +42,157 @@ class BidsDatasetDict(TypedDict):
+    def filter(
+        self, *, regex_search: bool = False, **filters: str | Sequence[str]
+    ) -> Self:
+        """Filter component based on provided entity filters
```
I agree. I wonder if we should have a position only argument `__spec` that takes either a str or list of str and applies that directly to the only entity. It defaults to `None`. Giving both a `__spec` and `filters` results in an error. 

This will keep a consistent `Expandable` API while providing more flexibility and user-friendliness. In the future, we may wish to add a `__spec` to the other `filter` methods to support more exotic filter mechanisms (lambdas, zip_lists, etc). I think in general the `Expandable` protocol won't need to know about these. This will let us tailor the `filter` method to the specific context

tkkuehn replied:
I was mainly referring to the docstring here, which seems close to identical for the one for `BidsComponent.filter`. That said, adding a position-only argument that defaults to `None` to provide a row-specific API seems like a reasonable idea to me!

kaitj reviewed (commented):
A couple of spelling things found, but pending resolution of Tristan's review, think this one may be good to go.

Re: your design question - what would be the alternative?

pvandyken commented:
> Re: your design question - what would be the alternative?

Not sure, guess I wanted to see if either of you would object to the status quo

tkkuehn commented:
> Should we store `.entity` as something else?

Sorry, forgot to specifically address this as part of my first review. I'd also be happy with `entity_name`, which I guess would slightly more precisely describe what part of the entity this attribute represents (i.e. the long-form name as opposed to the entity key or value). I think I'm fine with just calling it `entity` though.

pvandyken commented:
> Sorry, forgot to specifically address this as part of my first review. I'd also be happy with entity_name, which I guess would slightly more precisely describe what part of the entity this attribute represents (i.e. the long-form name as opposed to the entity key or value). I think I'm fine with just calling it entity though.

Okay, I'm just going to keep it then.

pvandyken commented:
I implemented the requested changes and the `__spec` argument on filter. In addition:

* I changed the filter api to `Iterable`. This makes it more flexible for use (not a breaking change), and in principle, safer in a snakemake environment where we don't have static type checking (the previous version ate iterables without filtering correctly). All the tests now use consumable iterators instead of lists.
* I went ahead and made `paths` in all the expand methods a position-only argument `__paths`. This is technically a breaking change, but I highly doubt anyone is referring to it via the keyword currently. My rationale is
  1. Accessing it via keyword makes it semantically confusing with `**wildcards`
  2. `paths` is technically a valid wildcard name, and we're taking up that namespace for no good reason. Of course, snakemake's `expand` does the same thing with `filepatterns` and `func`, and this is something we can address later, but in the meantime we should keep our own functions clean.

kaitj reviewed (approved):
lgtm pending merge conflict fix
